// furls/furls.js
// Generated by CoffeeScript 2.7.0
(function() {
  var Furls,
    indexOf = [].indexOf;

  Furls = class Furls {
    constructor() {
      this.inputs = [];
      //@inputsByName = {}  # unclear this data structure is worthwhile
      this.listeners = {
        loadURL: [],
        inputChange: [],
        stateChange: []
      };
      //# Coalesce multiple inputChange events into single stateChange event
      this.inputsChanged = {};
      this.on('inputChange', (input) => {
        //# Radio buttons might trigger multiple changes, but we only want to
        //# store the one that is now checked, to avoid overwriting that one
        //# with the same name.
        if (input.value == null) {
          return;
        }
        this.inputsChanged[input.name] = input;
        if (!this.microtask) {
          this.microtask = true;
          return this.queueMicrotask(() => {
            var key;
            this.microtask = false;
            if (((function() {
              var results1;
              results1 = [];
              for (key in this.inputsChanged) {
                results1.push(key);
              }
              return results1;
            }).call(this)).length === 0) {
              return;
            }
            this.trigger('stateChange', this.inputsChanged);
            return this.inputsChanged = {};
          });
        }
      });
    }

    queueMicrotask(task) {
      if ((typeof window !== "undefined" && window !== null ? window.queueMicrotask : void 0) != null) {
        return window.queueMicrotask(task);
      } else {
        return setTimeout(task, 0);
      }
    }

    on(event, listener) {
      this.listeners[event].push(listener);
      return this;
    }

    off(event, listener) {
      var i;
      if (0 <= (i = this.listeners[event].indexOf(listener))) {
        this.listeners[event].splice(i, 1);
      }
      return this;
    }

    trigger(event, ...args) {
      var j, len, listener, ref;
      ref = this.listeners[event];
      for (j = 0, len = ref.length; j < len; j++) {
        listener = ref[j];
        listener.call(this, ...args);
      }
      return this;
    }

    findInput(input, noCrash) {
      var inputObj, j, len, ref;
      //# Support various ways to specify input: internal input object,
      //# string ID of <input> element, or DOM object of <input> element.
      if (input.dom != null) {
        return input;
      } else {
        if (typeof input === 'string') { //# String ID
          input = document.getElementById(input);
        }
        ref = this.inputs;
        //# DOM object
        for (j = 0, len = ref.length; j < len; j++) {
          inputObj = ref[j];
          if (inputObj.dom === input) {
            return inputObj;
          }
        }
        if (noCrash) {
          return;
        }
        throw new Error(`Could not find input given ${input}`); //# Internal interface
      }
    }

    configInput(input, options) {
      var key, value;
      input = this.findInput(input);
      if (options != null) {
        for (key in options) {
          value = options[key];
          input[key] = value;
        }
      }
      return this;
    }

    
      //# Use .checked for checkboxes and radio buttons, .value for other inputs.
    //# Radio buttons use `undefined` to denote "not checked", to avoid overwriting
    //# the correct value from the checked button.
    //# Automatically parse value for type=number.
    //# <select> uses an array of values if `multiple` is set, value otherwise.
    getInputValue(input) {
      var j, len, option, ref, results1;
      switch (input.type) {
        case 'radio':
          if (input.dom.checked) {
            return input.dom.value;
          }
          break;
        case 'checkbox':
          return input.dom.checked;
        case 'number':
        case 'range':
          return parseFloat(input.dom.value);
        case 'select':
          if (input.dom.multiple) {
            ref = input.dom.selectedOptions;
            results1 = [];
            for (j = 0, len = ref.length; j < len; j++) {
              option = ref[j];
              results1.push(option.value);
            }
            return results1;
          } else {
            return input.dom.value;
          }
          break;
        default:
          return input.dom.value;
      }
    }

    getInputDefaultValue(input) {
      var j, len, multiple, option, ref;
      switch (input.type) {
        case 'radio':
          if (input.dom.defaultChecked) {
            // dom.value only works when dom.checked is true
            return input.dom.getAttribute('value');
          }
          break;
        case 'checkbox':
          return input.dom.defaultChecked;
        case 'number':
        case 'range':
          return parseFloat(input.dom.defaultValue);
        case 'select':
          multiple = input.dom.multiple;
          ref = input.dom.options;
          for (j = 0, len = ref.length; j < len; j++) {
            option = ref[j];
            if (!option.defaultSelected) {
              continue;
            }
            if (multiple) {
              option.value;
            } else {
              return option.value;
            }
          }
          break;
        default:
          return input.dom.defaultValue;
      }
    }

    setInputValue(dom, value) {
      var j, k, len, len1, option, ref, ref1, ref2, results1, results2;
      switch (dom.type) {
        case 'radio':
          return dom.checked = value === dom.getAttribute('value');
        case 'checkbox':
          //# Convert to Boolean checked for checkboxes
          switch (value) {
            case '1':
            case 'true':
            case true:
              value = true;
              break;
            case '0':
            case 'false':
            case false:
              value = false;
              break;
            default:
              value = !!value;
          }
          return dom.checked = value;
        case 'select':
          if (input.dom.multiple) {
            if (!Array.isArray(value)) {
              value = [value];
            }
            ref = input.dom.options;
            results1 = [];
            for (j = 0, len = ref.length; j < len; j++) {
              option = ref[j];
              results1.push(option.selected = (ref1 = option.value, indexOf.call(value, ref1) >= 0));
            }
            return results1;
          } else {
            ref2 = input.dom.options;
            results2 = [];
            for (k = 0, len1 = ref2.length; k < len1; k++) {
              option = ref2[k];
              if (option.value === value) {
                option.selected = true;
                break;
              } else {
                results2.push(void 0);
              }
            }
            return results2;
          }
          break;
        default:
          return dom.value = value;
      }
    }

    maybeChange(input, recurse = true, trigger = true) {
      var input2, j, len, ref, value;
      input = this.findInput(input);
      if (input.value !== (value = this.getInputValue(input))) {
        input.oldValue = input.value;
        input.value = value;
        if (trigger) {
          this.trigger('inputChange', input);
        }
        //# Auto-trigger change of all inputs with same name: radio buttons get
        //# events on the clicked button, but not on all the unset buttons.
        if (recurse) {
          ref = this.inputs;
          //for input2 in @inputsByName[input.name] when input != input2
          for (j = 0, len = ref.length; j < len; j++) {
            input2 = ref[j];
            if (input2.name === input.name && input !== input2) {
              input2.lastEvent = input.lastEvent;
              this.maybeChange(input2, false);
            }
          }
        }
      }
      if (this.isCommitChange(input)) {
        input.committed = input.value;
      }
      return this;
    }

    get(input) {
      return this.findInput(input).value;
    }

    set(input, value) {
      input = this.findInput(input);
      input.lastEvent = 'set';
      this.setInputValue(input.dom, value);
      return this.maybeChange(input);
    }

    //# <input>s and <textarea>s should trigger 'input' events during every change,
    //# (according to HTML5), and 'change' events when the change is "committed"
    //# (for text fields, when losing focus).  In some browsers, checkboxes and
    //# radio buttons don't trigger 'input', but they immediately trigger 'change'.
    //# So check for both, and just ignore the event if nothing changed.
    getInputEvents(input) {
      return ['input', 'change'];
    }

    //# Detect whether the last change to an input is a "commit":
    //# a 'change' event instead of an 'input' event, on inputs supporting both.
    //# (True for inputs not supporting both, and string events like "set".)
    isCommitChange(input) {
      var ref, ref1;
      return ((ref = input.type) === 'select' || ref === 'radio' || ref === 'checkbox' || ref === 'button' || ref === 'submit' || ref === 'reset') || typeof input.lastEvent === 'string' || ((ref1 = input.lastEvent) != null ? ref1.type : void 0) === 'change';
    }

    addInput(input, options) {
      var event, listener, ref;
      if (this.findInput(input, true)) {
        return this.configInput(input, options);
      }
      if (typeof input === 'string') {
        input = {
          id: input
        };
      } else if (input instanceof HTMLElement) {
        input = {
          dom: input
        };
      }
      if (input.dom == null) {
        input.dom = document.getElementById(input.id);
      }
      if (input.id == null) {
        input.id = input.dom.getAttribute('id');
      }
      if (input.type == null) {
        input.type = (function() {
          switch (input.dom.tagName.toLowerCase()) {
            case 'textarea':
              return 'textarea';
            case 'select':
              return 'select';
            case 'input':
              return input.dom.getAttribute('type').toLowerCase();
          }
        })();
      }
      if (input.name == null) {
        input.name = (ref = input.dom.getAttribute('name')) != null ? ref : input.id;
      }
      if (input.defaultValue == null) {
        input.defaultValue = this.getInputDefaultValue(input);
      }
      input.value = input.committed = this.getInputValue(input);
      this.configInput(input, options);
      this.inputs.push(input);
      //(@inputsByName[input.name] ?= []).push input
      input.listeners = (function() {
        var j, len, ref1, results1;
        ref1 = this.getInputEvents(input);
        results1 = [];
        for (j = 0, len = ref1.length; j < len; j++) {
          event = ref1[j];
          input.dom.addEventListener(event, listener = (e) => {
            input.lastEvent = e;
            return this.maybeChange(input);
          });
          results1.push(listener);
        }
        return results1;
      }).call(this);
      return this;
    }

    addInputs(selector = 'input, select, textarea') {
      var input, j, len;
      if (typeof selector === 'string') {
        selector = document.querySelectorAll(selector);
      }
      for (j = 0, len = selector.length; j < len; j++) {
        input = selector[j];
        this.addInput(input);
      }
      return this;
    }

    removeInput(input) {
      var event, i, j, k, len, len1, ref, ref1, target;
      input = this.findInput(input);
      ref = this.getInputEvents(input);
      for (i = j = 0, len = ref.length; j < len; i = ++j) {
        event = ref[i];
        input.dom.removeEventListener(event, input.listeners[i]);
      }
      if ((this._syncClass != null) && (target.value != null)) {
        ref1 = this._syncClass.selector;
        for (k = 0, len1 = ref1.length; k < len1; k++) {
          target = ref1[k];
          target.classList.remove(`${this._syncClass.prefix}${input.name}-${input.value}`);
        }
      }
      return this;
    }

    removeInputs(selector) {
      var input;
      if (typeof selector === 'string') {
        selector = document.querySelectorAll(selector);
      } else if (!Array.isArray(selector)) {
        selector = [selector];
      }
      this.inputs = (function() {
        var j, len, ref, ref1, results1;
        ref = this.inputs;
        results1 = [];
        for (j = 0, len = ref.length; j < len; j++) {
          input = ref[j];
          if (indexOf.call(selector, input) >= 0 || (ref1 = input.dom, indexOf.call(selector, ref1) >= 0)) {
            this.removeInput(input);
            continue;
          } else {
            results1.push(input);
          }
        }
        return results1;
      }).call(this);
      return this;
    }

    clearInputs() {
      return this.removeInputs(this.inputs);
    }

    getState() {
      var input, j, len, ref, state, value;
      state = {};
      ref = this.inputs;
      for (j = 0, len = ref.length; j < len; j++) {
        input = ref[j];
        value = this.getInputValue(input);
        //# Avoid overwriting the correct value for radio buttons sharing a name
        if (value != null) {
          state[input.name] = value;
        }
      }
      return state;
    }

    getSearch() {
      var input, search, value;
      search = ((function() {
        var j, len, ref, results1;
        ref = this.inputs;
        results1 = [];
        for (j = 0, len = ref.length; j < len; j++) {
          input = ref[j];
          value = this.getInputValue(input);
          if (value === input.defaultValue) {
            //# Don't store default values
            continue;
          }
          if (value == null) {
            //# Don't store off radio buttons; just need the "on" one
            continue;
          }
          if (input.encode != null) {
            //# Custom encoding
            value = input.encode(value);
          }
          //# Stringify booleans for checkboxes
          switch (value) {
            case true:
              value = '1';
              break;
            case false:
              value = '0';
          }
          results1.push(`${input.name}=${encodeURIComponent(value).replace(/%20/g, '+')}`);
        }
        return results1;
      }).call(this)).join('&');
      if (search) {
        search = `?${search}`;
      }
      return search;
    }

    getRelativeURL() {
      return `${window.location.pathname}${this.getSearch()}`;
    }

    //# Based on jolly.exe's code from http://stackoverflow.com/questions/901115/how-can-i-get-query-string-values-in-javascript
    getParameterByName(name, search = window.location.search) {
      var regex, results;
      //# https://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex
      name = name.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
      regex = new RegExp(`[\\?&]${name}=([^&#]*)`);
      results = regex.exec(search);
      if (results == null) {
        return null;
      }
      return decodeURIComponent(results[1].replace(/\+/g, " "));
    }

    loadURL(url = window.location, trigger = true) {
      var customDecode, input, j, k, l, len, len1, len2, len3, m, ref, ref1, ref2, ref3, ref4, ref5, search, value;
      this.loading = true;
      if (url.search != null) {
        search = url.search;
      } else if (url[0] === '?') {
        search = url;
      } else {
        search = (ref = (ref1 = /\?.*$/.exec(url)) != null ? ref1[0] : void 0) != null ? ref : '';
      }
      ref2 = this.inputs;
      //# Reset all inputs to defaults before loading new values, because we
      //# only put deviation from defaults in the URL.  This needs to be in a
      //# separate stage because of checkboxes.
      for (j = 0, len = ref2.length; j < len; j++) {
        input = ref2[j];
        this.setInputValue(input.dom, input.defaultValue);
        input.lastEvent = 'load';
      }
      ref3 = [false, true];
      //# Do custom decoding in a separate phase after inputs without
      //# custom decoding, so that custom decoding can depend on those inputs.
      for (k = 0, len1 = ref3.length; k < len1; k++) {
        customDecode = ref3[k];
        ref4 = this.inputs;
        for (l = 0, len2 = ref4.length; l < len2; l++) {
          input = ref4[l];
          if (!((input.decode != null) === customDecode)) {
            continue;
          }
          value = this.getParameterByName(input.name, search);
          if (value != null) {
            if (input.decode != null) {
              value = input.decode(value);
            }
          } else {
            value = input.defaultValue;
          }
          this.setInputValue(input.dom, value);
          //# Update value and oldValue, and optionally trigger inputChange event
          //# which eventually triggers a stateChange event.
          //# Don't recurse on identically named inputs, as we process all inputs.
          this.maybeChange(input, false, trigger);
        }
      }
      ref5 = this.inputs;
      for (m = 0, len3 = ref5.length; m < len3; m++) {
        input = ref5[m];
        //# Don't recurse on identically named inputs, as we process all inputs.
        this.maybeChange(input, false, trigger);
      }
      this.trigger('loadURL', search);
      //# Schedule after possibly triggered stateChange event.
      this.queueMicrotask(() => {
        return this.loading = false;
      });
      return this;
    }

    setURL(history = 'push', force) {
      var search;
      search = this.getSearch();
      if (force || search !== window.location.search) {
        window.history[history + 'State'](null, 'furls', `${window.location.pathname}${search}`);
      }
      return this;
    }

    replaceState(force) {
      return this.setURL('replace', force);
    }

    pushState(force) {
      return this.setURL('push', force);
    }

    syncState(history = 'auto', loadNow = true) {
      var input, j, len, ref;
      this.on('stateChange', (changed) => {
        var input, minor, name, ref;
        if (this.loading) {
          return;
        }
        //# Check whether all changes are minor, and force 'replace' in that case.
        minor = true;
        for (name in changed) {
          input = changed[name];
          if (!(input.minor || (history === 'auto' && ((ref = input.committed) !== input.oldValue && ref !== input.value)))) {
            minor = false;
            break;
          }
        }
        if (minor) {
          return this.setURL('replace');
        } else if (history === 'auto') {
          return this.setURL('push');
        } else {
          return this.setURL(history);
        }
      });
      window.addEventListener('popstate', () => {
        return this.loadURL();
      });
      //# On initial load, treat as transition from undefined values to defaults.
      if (loadNow) {
        ref = this.inputs;
        for (j = 0, len = ref.length; j < len; j++) {
          input = ref[j];
          input.value = void 0;
        }
        this.loadURL();
      }
      return this;
    }

    
      //# Which types have discrete values like `true` and `false`, and thus are
    //# appropriate for classes.
    discreteValue(input) {
      var ref;
      return (ref = input.type) === 'checkbox' || ref === 'radio';
    }

    syncClass(selector = [document.documentElement], prefix = '', updateNow = true) {
      var input, j, k, len, len1, ref, target;
      if (typeof selector === 'string') {
        selector = document.querySelectorAll(selector);
      } else if (!Array.isArray(selector)) {
        selector = [selector];
      }
      this.on('inputChange', (input) => {
        var j, len, results1, target;
        if (this.discreteValue(input)) {
          results1 = [];
          for (j = 0, len = selector.length; j < len; j++) {
            target = selector[j];
            if (input.oldValue != null) {
              target.classList.remove(`${prefix}${input.name}-${input.oldValue}`);
            }
            if (input.value != null) {
              results1.push(target.classList.add(`${prefix}${input.name}-${input.value}`));
            } else {
              results1.push(void 0);
            }
          }
          return results1;
        }
      });
      if (updateNow) {
        ref = this.inputs;
        for (j = 0, len = ref.length; j < len; j++) {
          input = ref[j];
          if (input.value != null) {
            if (this.discreteValue(input)) {
              for (k = 0, len1 = selector.length; k < len1; k++) {
                target = selector[k];
                target.classList.add(`${prefix}${input.name}-${input.value}`);
              }
            }
          }
        }
      }
      this._syncClass = {selector, prefix};
      return this;
    }

  };

  if (typeof module !== "undefined" && module !== null) {
    module.exports = Furls;
  }

  if (typeof window !== "undefined" && window !== null) {
    window.Furls = Furls;
  }

}).call(this);
